File: unity/Assets/_Project/Scripts/Core/State/MatchOutcome.cs
#nullable enable
using System.Collections.Generic;

namespace Pose.Core
{
    /// <summary>
    /// The result of a finished single-round match. <see cref="WinnerId"/> identifies
    /// the *triggering* player (the one who dominoed, or the lone holder of the
    /// lowest pip count); <see cref="WinningTeamId"/> identifies which team scores —
    /// the same as the winning player's team. Both are null only on a draw (a tied
    /// block end). <see cref="WinnerScore"/> is the sum of pips remaining in every
    /// losing player's hand; for a draw it is 0. Partner variants (introduced later
    /// in M1 step 3) may exclude teammates' pips from the score; the field's contract
    /// is just "the points the winning team earned this round".
    /// </summary>
    public sealed class MatchOutcome
    {
        public MatchEndReason Reason { get; }
        public PlayerId? WinnerId { get; }
        public TeamId? WinningTeamId { get; }
        public int WinnerScore { get; }
        public IReadOnlyDictionary<PlayerId, int> RemainingPips { get; }

        /// <summary>
        /// True when the win is a "key" — a both-ends lock-out (capicúa with no
        /// opponent holding either end). Worth the key bonus in a series; false
        /// for blocks, resigns and ordinary domino wins. See <see cref="KeyRule"/>.
        /// </summary>
        public bool IsKey { get; }

        public bool IsDraw => WinnerId == null;

        public MatchOutcome(
            MatchEndReason reason,
            PlayerId? winnerId,
            TeamId? winningTeamId,
            int winnerScore,
            IReadOnlyDictionary<PlayerId, int> remainingPips,
            bool isKey = false)
        {
            Reason = reason;
            WinnerId = winnerId;
            WinningTeamId = winningTeamId;
            WinnerScore = winnerScore;
            RemainingPips = remainingPips;
            IsKey = isKey;
        }
    }
}

File: functions/src/rules/matchOutcome.ts
import { PlayerId, TeamId } from './ids';

/** Why a round ended. Port of `Pose.Core.MatchEndReason`. */
export type MatchEndReason = 'domino' | 'blocked' | 'resigned';

/**
 * The result of a finished round. Port of `Pose.Core.MatchOutcome`.
 * `winnerId` / `winningTeamId` are null only on a draw (a tied block).
 * `winnerScore` is the pips summed across every losing hand (0 on a draw).
 * `remainingPips` maps each player to their leftover pip total.
 */
export interface MatchOutcome {
  readonly reason: MatchEndReason;
  readonly winnerId: PlayerId | null;
  readonly winningTeamId: TeamId | null;
  readonly winnerScore: number;
  /**
   * True when the win is a "key" — a both-ends lock-out (capicúa with no
   * opponent holding either end). Worth the key bonus in a series; false for
   * blocks, resigns and ordinary domino wins. Port of `Pose.Core.MatchOutcome.IsKey`.
   */
  readonly isKey: boolean;
  readonly remainingPips: ReadonlyMap<PlayerId, number>;
}

export function isDraw(outcome: MatchOutcome): boolean {
  return outcome.winnerId === null;
}
